File: 프로그래머스/기능개발.py
# def solution(progresses, speeds):
#     answer = []
#     days = 0  # 날짜 세기
#     cnt = 0  # 완료된 기능
#
#     while len(progresses) > 0:
#         # 기능의 진행상황과 그 동안 지난 날짜만큼의 speed를 구해서 더하기
#         if (progresses[0] + days * speeds[0]) >= 100:
#             # 완료되면 리스트에서 제거
#             progresses.pop(0)
#             speeds.pop(0)
#             # 완료된 기능 수
#             cnt += 1
#         else:
#             # 만약 완료된 기능이 있다면 answer에 더해주고 0으로 초기화
#             if cnt > 0:
#                 answer.append(cnt)
#                 cnt = 0
#             # 완료된 기능이 없으면 days 추가
#             else:
#                 days += 1
#     answer.append(cnt)
#     return answer
'''
progresses	                  speeds	        return
[93, 30, 55]	            [1, 30, 5]	        [2, 1]
[95, 90, 99, 99, 80, 99]	[1, 1, 1, 1, 1, 1]	[1, 3, 2]
'''


def solution(progresses, speeds):
    answer = []
    # 날짜 세기
    day = 0
    # 완료된 기능
    cnt = 0
    while len(progresses) > 0:  # 리스트의 요소가 있을때까지
        if (progresses[0] + day * speeds[0]) >= 100:
            # 100%라면 리스트에서 제거
            progresses.pop(0)
            speeds.pop(0)
            # 완료된 건수
            cnt += 1
        else:
            # 만약 완료된 기능이 있다면 answer에 더해주고 0으로 초기화
            if cnt > 0:
                answer.append(cnt)
                cnt = 0
            else:
                day += 1
    answer.append(cnt)
    return answer

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        // 값 수정이 가능하게끔 arraylist로
        List<Integer> ans = new ArrayList<>();
    
        // 배포 가능한 날짜
        List<Integer> upload = new ArrayList<>();
        
        for(int i=0; i<progresses.length; i++){
            upload.add((int)Math.ceil((100.0-progresses[i])/speeds[i]));
        }
        
        // 앞의 기능
        int pre = upload.get(0);
        int cnt = 1; // 함께 배포하는 작업 수
        
        for(int i=1; i<upload.size(); i++){
            // 앞의 배포 날짜가 더 크다면, 배포 날짜에 함께 배포한다
            if(upload.get(i) <= pre){
                cnt ++;
            }else{
                ans.add(cnt);
                cnt = 1;
                pre= upload.get(i); // 앞의 날짜 갱신
            }
        }
        ans.add(cnt); // 마지막 그룹까지 추가
        
        int[] answer = new int[ans.size()];
        
        for(int i=0; i<answer.length; i++){
            answer[i]=ans.get(i);
        }
        
        return answer;
    }
}